kvbc/src/v4blockchain/v4_blockchain.cpp

```diff
@@ -56,15 +56,44 @@ BlockId KeyValueBlockchain::add(categorization::Updates&& updates) {
     latest_keys_.addBlockKeys(updates, block_id, write_batch);
   }
   native_client_->write(std::move(write_batch));
-  block_chain_.setLastReachable(block_id);
+  block_chain_.setBlockId(block_id);
   if (sequence_number > 0) setLastBlockSequenceNumber(sequence_number);
   return block_id;
 }
 
+BlockId KeyValueBlockchain::deleteBlocksUntil(BlockId until) {
+  auto scoped = v4blockchain::detail::ScopedDuration{"deleteBlocksUntil"};
+  return block_chain_.deleteBlocksUntil(until);
+}
+
+void KeyValueBlockchain::deleteGenesisBlock() {
+  auto scoped = v4blockchain::detail::ScopedDuration{"deleteGenesisBlock"};
+  block_chain_.deleteGenesisBlock();
+}
+
+void KeyValueBlockchain::deleteLastReachableBlock() {
+  // validate conditions
+  auto last_reachable_id = block_chain_.getLastReachable();
+  auto genesis_id = block_chain_.getGenesisBlockId();
+  ConcordAssertLT(genesis_id, last_reachable_id);
+  // get block updates
+  auto write_batch = native_client_->getBatch();
+  auto updates = block_chain_.getBlockUpdates(last_reachable_id);
+  ConcordAssert(updates.has_value());
+  // revert from latest keys
+  latest_keys_.revertLastBlockKeys(*updates, last_reachable_id, write_batch);
```

Comment:
Because a when adding a block we:

1. add it as a block to the block CF
2. add its keys to the latest CF.

If we delete it, we want to revert the blockchain to the exact state as it was before the addition.